src/main/java/sx/blah/discord/util/MessageHistoryBuilder.java

```diff
@@ -0,0 +1,429 @@
+ *     GNU Lesser General Public License for more details.
+ *
+ *     You should have received a copy of the GNU Lesser General Public License
+ *     along with Discord4J.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package sx.blah.discord.util;
+
+import sx.blah.discord.handle.impl.obj.Channel;
+import sx.blah.discord.handle.obj.IChannel;
+import sx.blah.discord.handle.obj.IMessage;
+
+import java.time.LocalDateTime;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+
+import static sx.blah.discord.handle.impl.obj.Channel.MESSAGE_CHUNK_COUNT;
+
+/**
+ * This is a utility class for building message histories in a readable, easy-to-understand fashion. It can be built
+ * into a {@link MessageHistory} object, a {@link Stream} of {@link IMessage} objects, or an {@link Iterator} of
+ * {@link IMessage} objects. It can also be directly used in a for-each loop without being built.
+ *
+ * @see MessageHistory
+ * @see MessageHistoryIterator
+ */
+public class MessageHistoryBuilder implements Iterable<IMessage> {
+
+	private Channel channel;
+
+	// User provided info
+	private IMessage uStart; // Null if not given
+	private LocalDateTime uStartTime; // null signifies most recent
+	private boolean uIncludeStart;
+	private IMessage uEnd; // Null if not given
+	private LocalDateTime uEndTime; // null signifies channel start
+	private boolean uIncludeEnd;
+	private int count = -1; // < 0 signifies no limit
+
+	// Used for builds
+	private List<IMessage> cached;
+	private List<IMessage> messages;
+	private MessageHistoryRange range;
+	private int currentCount;
+
+	public MessageHistoryBuilder(IChannel channel) {
+		this.channel = (Channel) channel;
+	}
+
+	/**
+	 * Specifies the message to start at (inclusive).
+	 *
+	 * @param startID the message ID to start at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder startAt(long startID) {
+		this.uStart = channel.getMessageByID(startID);
+		if (uStart == null)
+			throw new IllegalArgumentException("Message " + Long.toUnsignedString(startID) + " does not exist in channel " + channel.getStringID() + "!");
+		this.uStartTime = uStart.getTimestamp();
+		this.uIncludeStart = true;
+		return this;
+	}
+
+	/**
+	 * Specifies the message to start at (inclusive).
+	 *
+	 * @param msg the message to start at
+	 * @return this, for chaining
+	 */
+	public MessageHistoryBuilder startAt(IMessage msg) {
+		if (!channel.equals(msg.getChannel()))
+			throw new IllegalArgumentException("Message " + msg.getStringID() + " does not exist in channel " + channel.getStringID() + "!");
+		this.uStart = msg;
+		this.uStartTime = uStart.getTimestamp();
+		this.uIncludeStart = true;
+		return this;
+	}
+
+	/**
+	 * Specifies the time to start at (inclusive).
+	 * NOTE: Using this method to set a start time chronologically before endAt, and using withMaxCount > 0, may
```

Comment:
It is only allowed because I wanted to allow "backwards" ranges (where the start is chronologically before the end- I.e. getting channel history from the first message in the channel to now). This is just an inconvenient artifact of the combination of worst case scenarios for building-
If a start time rather than message is provided, it has to find the first message, and in the case of backwards ranges, this entails going through all the target messages until the start is hit.
When count is unlimited it just pretends like the end is the start which works fine, but otherwise it has to find the start message so it can count backwards.
Theoretically I could instead still pretend like the end is the start, and then fetch the entire requested range and then reverse it and take only the first `limit` messages, but this comes with the danger of unexpectedly high memory usage storing the possibly very large number of messages just to get 5 messages at the end of the message history.